src/platform/nm-platform.c

```diff
@@ -4138,11 +4138,12 @@ nm_platform_ip4_address_sync (NMPlatform *self,
 		}
 	}
 	ip4_addr_subnets_destroy_index (plat_subnets, plat_addresses);
-	ip4_addr_subnets_destroy_index (known_subnets, known_addresses);
 
 	if (!known_addresses)
 		return TRUE;
 
+	ip4_addr_subnets_destroy_index (known_subnets, known_addresses);
```

Comment:
This warning is a false positive.

We can merge this commit, if only to shut up cppcheck.

But could you reword the commit message? I think there is no "possible" null pointer dereference.

And in the git-commit message, please always leave an empty line after the first one. `git log --oneline` will unwrap such line continuations, which is probably not intended here. Same with the parent commit.

Reply:
function `ip4_addr_subnets_destroy_index` checks first argument against NULL, but does not check second one.

if known_addresses is NULL, so it will be dereferenced inside `ip4_addr_subnets_destroy_index`

moving `if(!known_addresses)` earlier than calling `ip4_addr_subnets_destroy_index` eliminates that situation.

I do not see it is being false positive.

Reply:
`@known_subnets` is only not NULL, if `@known_addresses` is not NULL. You can easily see that by looking at `nm_platform_ip4_address_sync()`.

Reply:
there're couple of things

1) checking whether known_addresses is NULL or not
2) passing known_addresses to ip4_addr_subnets_destroy_index which dereferences known_addresses

either comparision should not happen at all, or it should happen earlier than dereference

Reply:
The point is that the commit message claims there is a bug there (NULL pointer dereference). That would be a serious thing. Commit messages are important, and it should not make incorrect claims about the seriousness. I only complain here about the commit message!

> 2. passing known_addresses to ip4_addr_subnets_destroy_index which dereferences known_addresses

`known_addresses` is only dereferences if `known_subnets` is not NULL, and `known_subnets` is only not NULL if `known_addresses` is not NULL. The code has no possible NULL pointer dereference. 

Yes, the logic there might be improved (for the human reader, and not to confuse cppcheck). Not confusing tools (like cppcheck) and getting rid of false positives is important. Readable code is also important. So no argument there.

But note that while your patch makes it easier to see that `@known_addresses` is never dereferenced with NULL, it now might appear that `@known_subnets` gets leaked. Of course, with and without the patch neither happens, it's just not very obvious.

What would make it clear are assertions:

```diff
diff --git i/src/platform/nm-platform.c w/src/platform/nm-platform.c
index 4ec161adc3e4..754f2b1f64ef 100644
--- i/src/platform/nm-platform.c
+++ w/src/platform/nm-platform.c
@@ -3905,10 +3905,12 @@ ip4_addr_subnets_destroy_index (GHashTable *subnets, const GPtrArray *addresses)
     gpointer p;
 
     if (!subnets)
          return;
 
+    nm_assert (addresses);
+
     g_hash_table_iter_init (&iter, subnets);
     while (g_hash_table_iter_next (&iter, NULL, &p)) {
          if (!ip4_addr_subnets_is_plain_address (addresses, p))
               g_ptr_array_free ((GPtrArray *) p, TRUE);
     }
```

Assertions with `nm_assert()` are not compiled in production code, but enabled during tests. They can be used liberally.

Reply:
well, it is not clear about `nm_assert`, what will happen in production then ?